File: adventofcode/2019/d4/d4.jl
function Counter(iter::String)::Dict{Char,Int64}
    d = Dict()
    for elem in iter
        d[elem] = get(d, elem, 0) + 1
    end
    return d
end

function valid(code::Int64, pred::Function)
    scode = string(code)
    if !all(b >= a for (a, b) in zip(scode[1:end - 1], scode[2:end]))
        return false
    end
    # 599922 => ('5', '9', '2'), (1, 3, 2)
    c = Counter(scode)
    digits, counts = collect(keys(c)), collect(values(c))
    return pred(counts)
end

a, b = 284639, 748759
# p1
println(sum(valid(code, counts->any(count > 1 for count in counts))
          for code in a:b))
# p2
println(sum(valid(code, counts->any(count == 2 for count in counts)) for code in a:b))

File: adventofcode/2019/d4/d4.rs
use std::collections::HashMap;

fn counter(iter: &str) -> HashMap<char, usize> {
  let mut d = HashMap::new();
  for elem in iter.chars() {
    let count = d.entry(elem).or_insert(0);
    *count += 1;
  }
  d
}

fn valid<F>(code: usize, pred: F) -> bool
where
  F: Fn(&Vec<&usize>) -> bool,
{
  let scode = code.to_string();
  if !scode[..]
    .chars()
    .zip(scode[1..].chars())
    .all(|(a, b)| b >= a)
  {
    return false;
  }
  let c = counter(&scode);
  let counts: Vec<&usize> = c.values().collect();
  pred(&counts)
}

fn main() {
  let a = 284639;
  let b = 748759;

  println!(
    "{}",
    (a..b)
      .map(|n| valid(n, |counts| counts.iter().any(|&c| *c > 1)))
      .map(|b| b as usize)
      .sum::<usize>()
  );

  println!(
    "{}",
    (a..b)
      .map(|n| valid(n, |counts| counts.iter().any(|&c| *c == 2)))
      .map(|b| b as usize)
      .sum::<usize>()
  );
}
